tests/util/Component.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+
+const assert = require('assert');
+const eslint = require('eslint');
+const Components = require('../../lib/util/Components');
+
+const ruleTester = new eslint.RuleTester({
+  parserOptions: {
+    ecmaVersion: 2018,
+    sourceType: 'module',
+    ecmaFeatures: {
+      jsx: true,
+    },
+  },
+});
+
+describe('Components', () => {
+  describe('static detect', () => {
+    function makeTest(name, test, done) {
+      return it(name, () => {
+        const rule = Components.detect((context, components, util) => ({
+          'Program:exit'(node) {
+            done(name, context, components, util, node);
+          },
+        }));
+
+        const tests = {
+          valid: [Object.assign(test, {
+            settings: {
+              react: {
+                version: 'detect',
+              },
+            },
+          })],
+          invalid: [],
+        };
+        ruleTester.run(name, rule, tests);
+      });
+    }
+
+    makeTest('Stateless Function Components', {
+      code: `import React from 'react'
+        function MyStatelessComponent() {
+          return <></>
+        }`,
+    }, (name, _context, components) => {
+      assert(components.length() === 1, `${name} should be detected component`);
+      Object.values(components.list()).forEach((component) => {
+        assert(
+          component.node.id.name === 'MyStatelessComponent',
+          `${name} should be named MyStatelessComponent`
+        );
+      });
+    });
+
+    makeTest('Class Components', {
+      code: `import React from 'react'
```

Comment:
all these tests should use `parsers.all` with `features` so we get maximum test coverage.